File: src/py/flwr/client/message_handler/message_handler.py
# Copyright 2022 Flower Labs GmbH. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Client-side message handler."""

from logging import WARN
from typing import Optional, Tuple, cast

from flwr.client.client import (
    maybe_call_evaluate,
    maybe_call_fit,
    maybe_call_get_parameters,
    maybe_call_get_properties,
)
from flwr.client.numpy_client import NumPyClient
from flwr.client.typing import ClientFnExt
from flwr.common import ConfigsRecord, Context, Message, Metadata, RecordSet, log
from flwr.common.constant import MessageType, MessageTypeLegacy
from flwr.common.recordset_compat import (
    evaluateres_to_recordset,
    fitres_to_recordset,
    getparametersres_to_recordset,
    getpropertiesres_to_recordset,
    recordset_to_evaluateins,
    recordset_to_fitins,
    recordset_to_getparametersins,
    recordset_to_getpropertiesins,
)
from flwr.proto.transport_pb2 import (  # pylint: disable=E0611
    ClientMessage,
    Reason,
    ServerMessage,
)


class UnexpectedServerMessage(Exception):
    """Exception indicating that the received message is unexpected."""


class UnknownServerMessage(Exception):
    """Exception indicating that the received message is unknown."""


def handle_control_message(message: Message) -> Tuple[Optional[Message], int]:
    """Handle control part of the incoming message.

    Parameters
    ----------
    message : Message
        The Message coming from the server, to be processed by the client.

    Returns
    -------
    message : Optional[Message]
        Message to be sent back to the server. If None, the client should
        continue to process messages from the server.
    sleep_duration : int
        Number of seconds that the client should disconnect from the server.
    """
    if message.metadata.message_type == "reconnect":
        # Retrieve ReconnectIns from recordset
        recordset = message.content
        seconds = cast(int, recordset.configs_records["config"]["seconds"])
        # Construct ReconnectIns and call _reconnect
        disconnect_msg, sleep_duration = _reconnect(
            ServerMessage.ReconnectIns(seconds=seconds)
        )
        # Store DisconnectRes in recordset
        reason = cast(int, disconnect_msg.disconnect_res.reason)
        recordset = RecordSet()
        recordset.configs_records["config"] = ConfigsRecord({"reason": reason})
        out_message = message.create_reply(recordset)
        # Return TaskRes and sleep duration
        return out_message, sleep_duration

    # Any other message
    return None, 0


def handle_legacy_message_from_msgtype(
    client_fn: ClientFnExt, message: Message, context: Context
) -> Message:
    """Handle legacy message in the inner most mod."""
    client = client_fn(message.metadata.dst_node_id, context.partition_id)

    # Check if NumPyClient is returend
    if isinstance(client, NumPyClient):
        client = client.to_client()
        log(
            WARN,
            "Deprecation Warning: The `client_fn` function must return an instance "
            "of `Client`, but an instance of `NumpyClient` was returned. "
            "Please use `NumPyClient.to_client()` method to convert it to `Client`.",
        )

    client.set_context(context)

    message_type = message.metadata.message_type

    # Handle GetPropertiesIns
    if message_type == MessageTypeLegacy.GET_PROPERTIES:
        get_properties_res = maybe_call_get_properties(
            client=client,
            get_properties_ins=recordset_to_getpropertiesins(message.content),
        )
        out_recordset = getpropertiesres_to_recordset(get_properties_res)
    # Handle GetParametersIns
    elif message_type == MessageTypeLegacy.GET_PARAMETERS:
        get_parameters_res = maybe_call_get_parameters(
            client=client,
            get_parameters_ins=recordset_to_getparametersins(message.content),
        )
        out_recordset = getparametersres_to_recordset(
            get_parameters_res, keep_input=False
        )
    # Handle FitIns
    elif message_type == MessageType.TRAIN:
        fit_res = maybe_call_fit(
            client=client,
            fit_ins=recordset_to_fitins(message.content, keep_input=True),
        )
        out_recordset = fitres_to_recordset(fit_res, keep_input=False)
    # Handle EvaluateIns
    elif message_type == MessageType.EVALUATE:
        evaluate_res = maybe_call_evaluate(
            client=client,
            evaluate_ins=recordset_to_evaluateins(message.content, keep_input=True),
        )
        out_recordset = evaluateres_to_recordset(evaluate_res)
    else:
        raise ValueError(f"Invalid message type: {message_type}")

    # Return Message
    return message.create_reply(out_recordset)


def _reconnect(
    reconnect_msg: ServerMessage.ReconnectIns,
) -> Tuple[ClientMessage, int]:
    # Determine the reason for sending DisconnectRes message
    reason = Reason.ACK
    sleep_duration = None
    if reconnect_msg.seconds is not None:
        reason = Reason.RECONNECT
        sleep_duration = reconnect_msg.seconds
    # Build DisconnectRes message
    disconnect_res = ClientMessage.DisconnectRes(reason=reason)
    return ClientMessage(disconnect_res=disconnect_res), sleep_duration


def validate_out_message(out_message: Message, in_message_metadata: Metadata) -> bool:
    """Validate the out message."""
    out_meta = out_message.metadata
    in_meta = in_message_metadata
    if (  # pylint: disable-next=too-many-boolean-expressions
        out_meta.run_id == in_meta.run_id
        and out_meta.message_id == ""  # This will be generated by the server
        and out_meta.src_node_id == in_meta.dst_node_id
        and out_meta.dst_node_id == in_meta.src_node_id
        and out_meta.reply_to_message == in_meta.message_id
        and out_meta.group_id == in_meta.group_id
        and out_meta.message_type == in_meta.message_type
        and out_meta.created_at > in_meta.created_at
    ):
        return True
    return False

File: src/ts/src/lib/message_handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateOutMessage = exports.handleLegacyMessageFromMsgType = exports.handleControlMessage = void 0;
// Copyright 2024 Flower Labs GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================
const client_1 = require("./client");
const transport_1 = require("../protos/flwr/proto/transport");
const recordset_1 = require("./recordset");
const recordset_compat_1 = require("./recordset_compat");
const reconnect = (reconnectMsg) => {
    let reason = transport_1.Reason.ACK;
    let sleepDuration = BigInt(0);
    if (reconnectMsg.seconds !== BigInt(0)) {
        reason = transport_1.Reason.RECONNECT;
        sleepDuration = reconnectMsg.seconds;
    }
    const disconnectRes = {
        reason,
    };
    return [
        {
            msg: { oneofKind: "disconnectRes", disconnectRes },
        },
        sleepDuration,
    ];
};
const handleControlMessage = (message) => {
    if (message.metadata.messageType === "reconnet") {
        let recordset = message.content;
        let seconds = recordset?.configsRecords["config"]["seconds"];
        const reconnectMsg = reconnect({ seconds: seconds });
        const disconnectMsg = reconnectMsg[0];
        const sleepDuration = reconnectMsg[1];
        if (disconnectMsg.msg.oneofKind === "disconnectRes") {
            let reason = disconnectMsg.msg.disconnectRes.reason;
            let recordset = new recordset_1.RecordSet();
            recordset.configsRecords["config"] = new recordset_1.ConfigsRecord({ reason: reason });
            let outMessage = message.createReply(recordset);
            return [outMessage, Number(sleepDuration)];
        }
    }
    return [null, 0];
};
exports.handleControlMessage = handleControlMessage;
const handleLegacyMessageFromMsgType = (client_fn, message, context) => {
    let client = client_fn(context);
    client.setContext(context);
    let messageType = message.metadata.messageType;
    let outRecordset;
    switch (messageType) {
        case "get_properties": {
            const getPropertiesRes = (0, client_1.maybeCallGetProperties)(client, (0, recordset_compat_1.recordSetToGetPropertiesIns)(message.content));
            outRecordset = (0, recordset_compat_1.getPropertiesResToRecordSet)(getPropertiesRes);
            break;
        }
        case "get_parameters": {
            const getParametersRes = (0, client_1.maybeCallGetParameters)(client, (0, recordset_compat_1.recordSetToGetParametersIns)(message.content));
            outRecordset = (0, recordset_compat_1.getParametersResToRecordSet)(getParametersRes, false);
            break;
        }
        case "train": {
            const fitRes = (0, client_1.maybeCallFit)(client, (0, recordset_compat_1.recordSetToFitIns)(message.content, true));
            outRecordset = (0, recordset_compat_1.fitResToRecordSet)(fitRes, false);
            break;
        }
        case "evaluate": {
            const evaluateRes = (0, client_1.maybeCallEvaluate)(client, (0, recordset_compat_1.recordSetToEvaluateIns)(message.content, true));
            outRecordset = (0, recordset_compat_1.evaluateResToRecordSet)(evaluateRes);
            break;
        }
        default: {
            throw `Invalid message type: ${messageType}`;
        }
    }
    return message.createReply(outRecordset);
};
exports.handleLegacyMessageFromMsgType = handleLegacyMessageFromMsgType;
const validateOutMessage = (outMessage, inMessageMetadata) => {
    let outMeta = outMessage.metadata;
    let inMeta = inMessageMetadata;
    if (outMeta.runId === inMeta.runId &&
        outMeta.messageId === "" &&
        outMeta.srcNodeId === inMeta.dstNodeId &&
        outMeta.dstNodeId === inMeta.srcNodeId &&
        outMeta.replyToMessage === inMeta.messageId &&
        outMeta.groupId === inMeta.groupId &&
        outMeta.messageType === inMeta.messageType &&
        outMeta.createdAt > inMeta.createdAt) {
        return true;
    }
    return false;
};
exports.validateOutMessage = validateOutMessage;
